src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp

```diff
@@ -435,52 +413,54 @@ MergeTreePrefetchedReadPool::ThreadsTasks MergeTreePrefetchedReadPool::createThr
             part->prefetch_step_marks = std::max<size_t>(
                 1, static_cast<size_t>(std::round(static_cast<double>(settings.filesystem_prefetch_step_bytes) / part->approx_size_of_mark)));
         }
-        else
-        {
-            /// Experimentally derived ratio.
-            part->prefetch_step_marks = static_cast<size_t>(
-                std::round(std::pow(std::max<size_t>(1, static_cast<size_t>(std::round(sum_marks / 1000))), double(1.5))));
-        }
 
         /// This limit is important to avoid spikes of slow aws getObject requests when parallelizing within one file.
         /// (The default is taken from here https://docs.aws.amazon.com/whitepapers/latest/s3-optimizing-performance-best-practices/use-byte-range-fetches.html).
         if (part->approx_size_of_mark
             && settings.filesystem_prefetch_min_bytes_for_single_read_task
             && part->approx_size_of_mark < settings.filesystem_prefetch_min_bytes_for_single_read_task)
         {
-
-            const size_t new_min_prefetch_step_marks = static_cast<size_t>(
+            const size_t min_prefetch_step_marks_by_total_cols = static_cast<size_t>(
                 std::ceil(static_cast<double>(settings.filesystem_prefetch_min_bytes_for_single_read_task) / part->approx_size_of_mark));
+            /// At least one task to start working on it right now and another one to prefetch in the meantime.
+            const size_t new_min_prefetch_step_marks = std::min<size_t>(min_prefetch_step_marks_by_total_cols, sum_marks / threads / 2);
             if (min_prefetch_step_marks < new_min_prefetch_step_marks)
             {
-                LOG_TEST(
-                    log, "Increasing min prefetch step from {} to {}", min_prefetch_step_marks, new_min_prefetch_step_marks);
+                LOG_DEBUG(log, "Increasing min prefetch step from {} to {}", min_prefetch_step_marks, new_min_prefetch_step_marks);
```

Comment:
changed to `debug` because on `test` level we now have literally tons of log messages per query